cobalt/user_agent_platform_info.cc

```diff
@@ -0,0 +1,602 @@
+// Copyright 2024 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cobalt/user_agent_platform_info.h"
+
+#include <map>
+#include <memory>
+#include "base/command_line.h"
+#include "base/logging.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_util.h"
+#include "base/strings/stringprintf.h"
+#include "starboard/common/system_property.h"
+#include "starboard/extension/platform_info.h"
+#include "starboard/system.h"
+#if SB_IS(EVERGREEN)
+#include "chrome/updater/util.h"
+#include "starboard/extension/installation_manager.h"
+#endif  // SB_IS(EVERGREEN)
+
+namespace cobalt {
+
+void GetUserAgentInputMap(
+    const std::string& user_agent_input,
+    std::map<std::string, std::string>& user_agent_input_map) {
+  struct state {
+    std::string field{""};
+    std::string value{""};
+    bool field_value_delimiter_found{false};
+
+    void reset() {
+      field.clear();
+      value.clear();
+      field_value_delimiter_found = false;
+    }
+  } current_state;
+
+  char escape_char = '\\';
```

Comment:
Can these variables here be const? Even better, constexpr?

Reply:
Actually this whole function looks suspiciously like a  
tokenizer. If so, consider (or leave a TODO) using //base  
helpers.

https://source.chromium.org/chromium/chromium/src/+/main:base/strings/string_split.h;l=111-115;drc=2e06ecfe68db301ab7c136a0e992a58ff8fca54c

Reply:
This helper function, borrowed from C25, is highly customized for extracting UA properties from the command line user_agent flag. It's more efficient than using base APIs, and I do not see it as necessary to rewrite it.